vdr/didstore/writer.go

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (C) 2022 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package didstore
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+)
+
+func writeEventList(tx stoabs.WriteTx, newEventList eventList, id did.DID) error {
+	// update MetaRefs
+	for i := range newEventList.Events {
+		newEventList.Events[i].MetaRef = fmt.Sprintf("%s%d", id.String(), i)
+	}
+
+	nelBytes, _ := json.Marshal(newEventList)
+	eventShelf := tx.GetShelfWriter(eventShelf)
+	err := eventShelf.Put(stoabs.BytesKey(id.String()), nelBytes)
+	if err != nil {
+		return fmt.Errorf("writeEventList: database error: %w", err)
+	}
+	return nil
+}
+
+func writeDocument(tx stoabs.WriteTx, didDocument did.Document, transaction Transaction) error {
+	var (
+		documentWriter stoabs.Writer
+		documentBytes  []byte
+	)
+	// transaction
+	transactionWriter := tx.GetShelfWriter(transactionIndexShelf)
+	err := transactionWriter.Put(stoabs.HashKey(transaction.Ref), []byte{0})
+	if err != nil {
+		return fmt.Errorf("writeDocument: database error on txRef write: %w", err)
+	}
+
+	// document
+	documentWriter = tx.GetShelfWriter(documentShelf)
+	documentBytes, err = json.Marshal(didDocument)
+	if err != nil {
+		return fmt.Errorf("writeDocument: marshalling DID Document: %w", err)
+	}
+	err = documentWriter.Put(stoabs.HashKey(transaction.PayloadHash), documentBytes)
+	if err != nil {
+		return fmt.Errorf("writeDocument: database error on document write: %w", err)
+	}
+	return nil
+}
+
+func writeLatest(tx stoabs.WriteTx, id did.DID, metadata documentMetadata) error {
+	// store latest
+	latestWriter := tx.GetShelfWriter(latestShelf)
+
+	// store updated eventList (return new list?)
+	mdID := fmt.Sprintf("%s%d", id.String(), metadata.Version)
+	err := latestWriter.Put(stoabs.BytesKey(id.String()), []byte(mdID))
+	if err != nil {
+		return fmt.Errorf("writeDocument: database error on latest write: %w", err)
```

Comment:
having these operations as prefix (`writeDocument`) make it very tightly-coupled and error-prone (should it ever change). I generally prefer adding in the error handler of the caller, if important;

```
err = writeDocument()
if err != nil {
  return fmt.Errorf("writeDocument failed: %w", err)
}
```